lib/heidrun/mapping_tools/marc.rb

```diff
@@ -0,0 +1,204 @@
+module Heidrun
+  module MappingTools
+    ##
+    # Static methods for evaluations and assignments that can be shared between
+    # mappings of MARC providers.
+    #
+    module MARC
+      require_relative 'marc/genre'
+      require_relative 'marc/dctype'
+      require_relative 'marc/dcformat'
+
+      module_function
+
+      ##
+      # Given options representing leader and controlfield values, return an
+      # array of applicable genre (edm:hasType) controlled-vocabulary terms
+      #
+      # Options are as follows, and are all Strings:
+      #   - leader:  value of MARC 'leader' element
+      #   - cf_007:  value of MARC 'controlfield' with 'tag' attribute '007'
+      #   - cf_008:  value of MARC 'controlfield' with 'tag' attribute '008'
+      #   - cf_970a:  TODO
+      #
+      # @param opts [Hash] Options, as outlined above
+      # @return [Array]
+      def genre(opts)
+        genres = []
+        args = [genres, opts]
+        # TODO: insert handling of optional cf_970a for Hathi, as the first
+        # evaluation.
+        Genre.assign_language(*args) || Genre.assign_musical_score(*args) \
+          || Genre.assign_manuscript(*args) || Genre.assign_maps(*args) \
+          || Genre.assign_projected(*args) || Genre.assign_two_d(*args) \
+          || Genre.assign_nonmusical_sound(*args) \
+          || Genre.assign_musical_sound(*args)
+        genres << 'Government Document' \
+          if Genre.government_document?(opts[:cf_008])
+        genres
+      end
+
+      def dctype(opts)
+        types = []
+        args = [types, opts]
+        DCType.assign_337a(*args)
+        DCType.assign_text(*args) \
+          || DCType.assign_still_and_moving_image(*args) \
+          || DCType.assign_sound(*args) \
+          || DCType.assign_physical_object(*args) \
+          || DCType.assign_collection(*args) \
+          || DCType.assign_interactive_rsrc(*args)
+        types
+      end
+
+      def dcformat(opts)
+        formats = []
+        DCFormat.assign_from_leader(formats, opts)
+        DCFormat.assign_from_cf007(formats, opts)
+        formats.compact
+      end
+
+      ##
+      # Return a lambda suitable for Array#select, that gives the XML element
+      # with a particular name and 'tag' attribute.
+      #
+      # @example
+      #   The XML element for control field 007 has a 'name' of 'controlfield'
+      #   and a 'tag' of '007', so:
+      #     r.node.children
+      #      .select(&name_tag_condition('controlfield', '007')).first
+      #   gives you the matching element.
+      #
+      # @param name [String] The element name, without the namespace
+      # @param tag  [String] The value of the element's 'tag' attribute
+      def name_tag_condition(name, tag)
```

Comment:
I'm wondering if `#name_tag_condition` is necessary - we could probably get by with `#name_tag_match_condition`...

Either way, I don't think this a blocker for now.